modules/library/jdbc/src/main/java/org/geotools/jdbc/JDBCInsertFeatureWriter.java

```diff
@@ -71,28 +88,65 @@ public void remove() throws IOException {
     }
 
     public void write() throws IOException {
+        if (++curBufferPos >= buffer.length) {
+            //buffer full => do the inserts
+            flush();
+        }
+    }
+
+    @Override
+    protected void cleanup() throws IOException {
+        try {
+            flush();
+        } finally {
+            for (int i = 0; i < buffer.length; i++) {
+                if (buffer[i] == null) {
+                    break;
+                }
+                buffer[i].close();
+                buffer[i] = null;
+            }
+            super.cleanup();
+        }
+    }
+
+    private void flush() throws IOException {
+        if(curBufferPos == 0) {
+            return;
+        }
         try {
             //do the insert
-            dataStore.insert(last, featureType, st.getConnection());
-            
-            //the datastore sets as userData, grab it and update the fid
-            String fid = (String) last.getUserData().get( "fid" );
-            last.setID( fid );
-            
-            ContentEntry entry = featureSource.getEntry();
-            ContentState state = entry.getState( this.tx );            
-            state.fireFeatureAdded( featureSource, last );
+            Collection<ResultSetFeature> features =
+                    Arrays.asList(Arrays.copyOfRange(buffer, 0, curBufferPos));
+            dataStore.insert(features, featureType, st.getConnection());
+
+            for (ResultSetFeature cur : features) {
+                //the datastore sets as userData, grab it and update the fid
+                final String fid = (String) cur.getUserData().get("fid");
+                cur.setID(fid);
+                final SimpleFeature orig = (SimpleFeature)cur.getUserData().get("_original_");
+                if (orig != null) {
+                    ((FeatureIdImpl)orig.getIdentifier()).setID(fid);
+                    orig.getUserData().putAll(cur.getUserData());
+                    orig.getUserData().remove("_original_");
+                }
+
+                final ContentEntry entry = featureSource.getEntry();
+                final ContentState state = entry.getState(this.tx);
+                state.fireFeatureAdded(featureSource, cur);
```

Comment:
I know that uDig pays attention to events, @jodygarnett , is having a delayed notification of the insert (e.g., after 100 write calls) going to cause issues?